common/djangoapps/student/models.py

```diff
@@ -823,10 +823,13 @@ def is_course_full(self, course):
 
     def users_enrolled_in(self, course_id):
         """Return a queryset of User for every user enrolled in the course."""
-        return User.objects.filter(
+        course_users = User.objects.filter(
             courseenrollment__course_id=course_id,
-            courseenrollment__is_active=True
+            courseenrollment__is_active=True,
         )
+        anon_course_users = [user for user in course_users if not UserProfile.has_registered(user)]
+        enrolled_course_users = course_users.exclude(id__in=[anon_user.id for anon_user in anon_course_users])
```

Comment:
Do we actually need this additional list comprehension? Could we boost performance by doing:

```python
anonymous_user_ids = [
    user.id
    for user in course_users
    if not UserProfile.has_registered(user)
]
enrolled_course_users = course_users.exclude(id__in=anonymous_user_ids)
```